geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,362 @@
+        final int batchSize) {
+      this.context = context;
+      this.region = region;
+      this.sender = sender;
+      this.maxRate = maxRate;
+      this.batchSize = batchSize;
+    }
+
+    @Override
+    public CliFunctionResult call() throws Exception {
+      return replicateRegion(context, region, sender, maxRate, batchSize);
+    }
+  }
+
+  private CliFunctionResult replicateRegion(FunctionContext<String[]> context, Region region,
+      GatewaySender sender, long maxRate, int batchSize) {
+    final Set<?> entries;
+    if (region instanceof PartitionedRegion && sender.isParallel()) {
+      entries =
+          ((PartitionedRegion) region).getDataStore().getAllLocalBucketRegions()
+              .stream()
+              .flatMap(br -> ((Set<?>) br.entrySet()).stream()).collect(Collectors.toSet());
+    } else {
+      entries = region.entrySet();
+    }
+
+    final InternalCache cache = (InternalCache) context.getCache();
+    int replicatedEntries = 0;
+    final long startTime = clock.millis();
+
+    Iterator<?> iter = entries.iterator();
+
+    while (iter.hasNext()) {
+      int batchIndex = 0;
+      List<GatewaySenderEventImpl> batch = new ArrayList<>();
+      while (iter.hasNext() && batchIndex < batchSize) {
+        GatewaySenderEventImpl event = createGatewaySenderEvent(cache, (InternalRegion) region,
+            sender, (Region.Entry) iter.next());
+        if (event != null) {
+          batch.add(event);
+          batchIndex++;
+        }
+      }
+      try {
+        sendBatch(sender,
+            replicatedEntries, maxRate, startTime, batch);
+        replicatedEntries += batch.size();
+      } catch (InterruptedException e) {
+        return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+            "Operation canceled after having replicated " + replicatedEntries + " entries");
+      }
+    }
+
+    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+        "Entries replicated: " + replicatedEntries);
+  }
+
+  private GatewaySenderEventImpl createGatewaySenderEvent(InternalCache cache,
+      InternalRegion region, GatewaySender sender, Region.Entry entry) {
+    final EntryEventImpl event;
+    if (region instanceof PartitionedRegion) {
+      event = createEventForPartitionedRegion(sender, cache, region, entry);
+    } else {
+      event = createEventForReplicatedRegion(cache, region, entry);
+    }
+    if (event == null) {
+      return null;
+    }
+    try {
+      return new GatewaySenderEventImpl(EnumListenerEvent.AFTER_CREATE, event, null, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  private void sendBatch(GatewaySender sender,
+      int replicatedEntries, long maxRate, long startTime,
+      List batch) throws InterruptedException {
+    ((AbstractGatewaySender) sender).getEventProcessor().getDispatcher().dispatchBatch(batch, false,
```

Comment:
I was thinking sending the batch and reading the ack would be completely outside the sender.

I was thinking you'd get your own dedicated connection or connections and execute the methods to send the batch and get the response.

Like:

- get the sender's pool
- get a connection using that pool (I think with acquireConnection())
- execute the methods in SenderProxy

I don't think you'd actually have to create a SenderProxy. Once you have the connection, you could just call:
```
GatewaySenderBatchOp.executeOn(connection, pool, events, batchId, removeFromQueueOnException, isRetry);
return GatewaySenderBatchOp.executeOn(connection, pool);
```

The first call sends the batch and the second reads the ack.

Or add a new method to do both of those in one call.

In the current implementation, the AckReaderThread reads an ack for a batch that it doesn't know about. I'm not sure if there would be issues with this.

Reply:
I tried to reuse code by using the `GatewaySenderEventProcessor` machinery but I guess you are right. There could be issues with batches the `AckReader` does not know about and it is probably better to separate the regular replication batches from these new ones.
I have followed your very detailed directions (very appreciated!). I do not know if we should care about the batchId which could collide with the current ones. Also, the error handling is not very sophisticated for now.